### PR TITLE
Add support for Gym

### PR DESCRIPTION
**Pull request by MarcCote**
This PR adds support for OpenAI's Gym environments.

Here are some of the functionalities provided by `textworld.gym.*`
- `register_game`: create a Gym's environment for playing a single TW game. You can request extra information which will be provided in the `infos` dictionary returned by `env.step` and `env.reset`.
- `register_games`: create a Gym's environment for playing a **several** TW games. Each time `env.reset` is called, a new game from the pool of games is started. You can request extra information which will be provided in the `infos` dictionary returned by `env.step` and `env.reset`.
- `make_batch`: create a Gym's environment to play a batch of games simultaneously. On the backend, individual TW game can be run in parallel using different processes (faster) or a single process (allow for debugging).

### CODE REVIEW

**Comment by MarcCote** [comment]
@tavianator I added the wrapper we discussed. I'm still not settled on its name. Any idea?
https://github.com/Microsoft/TextWorld/pull/90/files#diff-1f38498a11c06208d4a3b68780bf8c5aR10

File: textworld/envs/wrappers/filter.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    def __init__(self, request_infos) -> None:
+        """
+        Arguments:
+            request_infos:
+                Specify which additional information from the `GameState` object
+                should be available in the `infos` dictionary returned by
+                `env.reset()` and `env.step()`. Possible choices are:
+
+                 * `'description'`: text description of the current room,
```
Comment (tavianator) [inline textworld/envs/wrappers/filter.py]: An enum might be nice (https://docs.python.org/3/library/enum.html)

File: textworld/envs/wrappers/filter.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            if attr.startswith("extra:"):
+                infos[attr] = game_state.extras.get(attr.split(":")[-1])
+            else:
+                infos[attr] = getattr(game_state, attr)
+
+        return infos
+
+    def step(self, command: str) -> Tuple[str, float, bool, Mapping[str, Any]]:
+        game_state, score, done = super().step(command)
+        ob = game_state.feedback
```
Comment (tavianator) [inline textworld/envs/wrappers/filter.py]: Just curious, why "ob"?

File: textworld/envs/wrappers/filter.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        infos = {}
+        for attr in self.request_infos:
+            if attr.startswith("extra:"):
+                infos[attr] = game_state.extras.get(attr.split(":")[-1])
+            else:
+                infos[attr] = getattr(game_state, attr)
+
+        return infos
+
+    def step(self, command: str) -> Tuple[str, float, bool, Mapping[str, Any]]:
```
Comment (tavianator) [inline textworld/envs/wrappers/filter.py]: Maybe a dedicated type is better than four separate return values?

File: textworld/envs/wrappers/filter.py (hunk 1 above)
Comment (tavianator) [reply textworld/envs/wrappers/filter.py]: (The docs say `extras:`; the code says `extra:`.)

You could do something like

```python
from enum import Enum

class GameInfo(Enum):
    DESCRIPTION = "description"
    INVENTORY = "inventory"
    MAX_SCORE = "max_score"
    # etc.
```

But if there's an unlimited number of extras you might want access to, then this won't work.

Comment (MarcCote) [reply textworld/envs/wrappers/filter.py]: Right, the extras are unlimited.
Thanks for the typo "extra**s**".

Comment (MarcCote) [reply textworld/envs/wrappers/filter.py]: Ideally, you'd need a way of asking for which information is available.

Comment (MarcCote) [reply textworld/envs/wrappers/filter.py]: Also, to give you an idea of what extra can be. In the games for the competition, agents can request `extra:recipe` (content of the cookbook) or `extra:walkthrough` (one possible sequence of commands that leads to winning the game).

Comment (tavianator) [reply textworld/envs/wrappers/filter.py]: Yeah then this is probably as good as we can do.  Might be nice to avoid duplicating the docs a bunch though.

Comment (MarcCote) [reply textworld/envs/wrappers/filter.py]: Yes, I know. I find this annoying too and very difficult to maintain. :(

Comment (tavianator) [reply textworld/envs/wrappers/filter.py]: `EnvironmentOptions`?  That sounds great to me.

File: textworld/gym/envs/batch_env.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+
+class _ChildEnv:
+    """
+    Wrapper for an env in a child process.
+    """
+    def __init__(self, id):
+        self._pipe, child_pipe = mp.Pipe()
+        self._process = mp.Process(target=_child, args=(id, child_pipe))
+        self._process.daemon = True
```
Comment (MarcCote) [inline textworld/gym/envs/batch_env.py]: @tavianator I added this to make sure the children get killed if the parent process dies.

File: textworld/envs/wrappers/filter.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                 'entities', 'verbs', 'command_templates',
+                 'admissible_commands', 'intermediate_reward',
+                 'policy_commands',
+                 'extras']
+
+    def __init__(self, **kwargs):
+        #: bool: Text description of the current room, i.e. output of the
+        #:       `look` command.
+        #:       This information changes from one step to another.
+        self.description = kwargs.get("description")
```
Comment (tavianator) [inline textworld/envs/wrappers/filter.py]: Might be nice to do `.get("description", False)` if the type is documented as `bool`

Comment (MarcCote) [reply textworld/envs/wrappers/filter.py]: Done

**Pull request merged**